design/object-bucket/object-bucket.md

```diff
@@ -2,38 +2,46 @@
+- Enable cluster admin control over bucket creation and object store access via the Kubernetes API
 - Restrict users from creating buckets with their access keys outside the view of Kubernetes (e.g. via `s3cmd` CLI tool)
-- Utilize a familiar and deterministic pattern when injecting bucket connection information into workload environments.
-- Ensure the app Pod will not be started until the bucket has been created.
+- Utilize a familiar and deterministic pattern when injecting bucket connection information into workload environments
+- Ensure the consuming Pod will not be started until the bucket has been created
 
 ## Non-Goals
 
-- This proposal does not implement a generalized method for bucket provisioning; it will be specific to Rook-Ceph
-- This design does not attempt to implement a 1:1 model of Kubernetes PersistentVolumes and PersistentVolumeClaims.  Some shallow similarities will exist. 
+- This design does not describe a generalized bucket provisioner, only a standard API by which to request buckets.
+- Pods will not be capable of attaching OBCs like they can PVCs.  This design does not introduce changes to core Kubernetes APIs.
+- This design does not provide users a means of deleting buckets via the Kubernetes API so as to avoid accidental loss of data.  A Delete operation only deletes operator created API resources.
 - This design does not include the Swift interface implementation by Ceph Object.
-- This design does not provide users a means of deleting buckets via the Kubernetes API so as to avoid accidental loss of data.  A Delete operation only deletes the Kubernetes API objects.
 
 ## Requirements
 
-1. The Rook-Ceph operator should be the only object store accessor with bucket CREATE privileges.
-1. User access key privileges must only have _object_ PUT, GET, and DELETE rights for the bucket they have requested.
-1. Users should be provided credential and connection information in a Pod-attachable resource.
-1. Users should be able to reacquire bucket credential and connection information should the CephObjectBucketClaim or sub-resource be deleted.
+1. The Rook-Ceph operator must have a single access key for bucket creation and reuse this key for the lifetime of the object store.
+1. Users must have permissions to PUT/GET/DELETE objects in the requested bucket
+1. Users must have permissions to define the access policies of objects in the requested bucket
+1. Users must be provided credential and connection information in a Pod-consumable resource
+1. Users must be able to reacquire bucket credential and connection information should the ObjectBucketClaim or sub-resource be deleted
+1. Deleting ObjectBucketClaims and/or ObjectBuckets should not cause the deletion of a bucket to prevent accidental loss of data
+1. Bucket naming conventions should protect users from bucket name collisions in the object store by concatenating the `metadata.namespace` and `metadata.name` of the OBC.
```

Comment:
Is the bucket name merely a convention? I go back and forth on the idea of composing a bucket name on behalf of the user vs. letting the user specify the desired bucket name. The latter works better for brownfield, but is worse from the perspective of being  reasonably assured the desired bucket name will be unique with the object store (which can span namespaces).

Reply:
Using the namespace to compose the bucket name also protects users from inter-namespace name collisions for greenfield.

WRT brownfield, if we allow users to define a brownfield bucket, does this mean we're generating a new user access key for them?  If that's the case, users would not be protected from accidental, cross-namespace bucket access.  For instance, user A in namespace A has bucket "test".  User B in namespace B, not knowing a bucket is already named "test", creates an OBC defining that as their bucket name.  If we support brownfield, we set up a situation where we've allowed user B to access user A's "test" bucket.